core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1275,21 +1275,42 @@ int runTranslogRecovery(Engine engine, Translog.Snapshot snapshot) throws IOExce
         return opsRecovered;
     }
 
+    /** creates an empty index and translog and opens the engine **/
+    public void createIndexAndTranslog() throws IOException {
+        assert recoveryState.getRecoverySource().getType() == RecoverySource.Type.EMPTY_STORE;
+        assert shardRouting.primary() && shardRouting.isRelocationTarget() == false;
+        // note: these are set when recovering from the translog
+        final RecoveryState.Translog translogStats = recoveryState().getTranslog();
+        translogStats.totalOperations(0);
+        translogStats.totalOperationsOnStart(0);
+        globalCheckpointTracker.updateGlobalCheckpointOnReplica(SequenceNumbers.NO_OPS_PERFORMED, "index created");
```

Comment:
I think a nicer approach (can be a follow-up done by me) would be not to call `updateGlobalCheckpointOnReplica` here, but instead call 

```
globalCheckpointTracker.activatePrimaryMode(SequenceNumbers.NO_OPS_PERFORMED);
```

either here or in the IndexShard constructor (where we create the GlobalCheckpointTracker) when the recovery source is EMPTY_STORE.

Reply:
yeah, I thought we talked about it. I would prefer to unify on when we activate the primary shard - on post recovery or on shard started. The activation logic is the same for local_shards and snapshot. For local_store we need the in sync set to be updated, which may be done as part of the constructor. I personally prefer a shard to be full ready when in `POST_RECOVERY` so I'm +1 on moving it there.

Reply:
I would prefer not to call `updateGlobalCheckpointOnReplica` on the `GlobalCheckpointTracker` if the shard is a blessed primary. A shard that's created from snapshot / local_store / local_shards is by definition blessed from the master. It should just activate the tracker. The activation logic for a replica can be different than for a primary.